src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java

```diff
@@ -118,6 +116,96 @@ public synchronized void close() throws IOException {
             return;
         }
 
+        try {
+            stopGracefully();
+        } finally {
+            stopForcefully();
+        }
+    }
+
+    @SuppressFBWarnings("SWL_SLEEP_WITH_LOCK_HELD")
+    private void stopGracefully() {
+        if (!process.isAlive()) {
+            return;
+        }
+        int pid = ((ProcessImplForWin32) process).getPid();
+        boolean sentConsoleCtrlEvent = false;
+        synchronized (Kernel32.INSTANCE) {
+            // First, start a separate process that holds the console alive
+            // so that later gg can re-attach to this same console
+            Process holderProc;
+            try {
+                // Waits indefinitely for a keystroke
+                holderProc = new ProcessBuilder().command("cmd", "/C", "pause").start();
+            } catch (IOException e) {
+                logger.atError(STOP_GRACEFULLY_EVENT).cause(e)
+                        .log("Failed to start holder process. Cannot stop gracefully");
+                return;
+            }
+
+            Kernel32 k32 = Kernel32.INSTANCE;
+            try {
+                // Must detach from current console before attaching to another
+                if (!k32.FreeConsole()) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("FreeConsole error {}", k32.GetLastError());
+                }
+                // calling attachConsole right after a process is launched will fail with invalid handle error
+                // waiting a bit ensures that we can attach to a process that just got launched.
+                try {
+                    Thread.sleep(500);
+                } catch (InterruptedException ignored) {
+                }
+                // Attach to the console that's running the target process
+                if (!k32.AttachConsole(pid)) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("AttachConsole error {}", k32.GetLastError());
+                    return;
+                }
+                // Make gg ignore Ctrl-C temporarily so that we don't terminate ourselves as well
+                if (!Kernel32Ex.INSTANCE.SetConsoleCtrlHandler(null, true)) {
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("SetConsoleCtrlHandler add error {}", k32.GetLastError());
+                    return;
+                }
+                // Send Ctrl-C to all processes in the console
+                if (k32.GenerateConsoleCtrlEvent(Wincon.CTRL_C_EVENT, 0)) {
+                    sentConsoleCtrlEvent = true;
+                } else {
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("GenerateConsoleCtrlEvent error {}", k32.GetLastError());
+                }
+            } finally {
+                // Re-attach gg to original console and re-enable Ctrl-C
```

Comment:
These need to be guarded.

`finally` runs _after_ `return` so any of the early returns you have will still execute this finally which may not be desirable.

Reply:
The first thing we do in the try block is freeconsole(), we need to reattach the console irrespective of where we return.